CleverRaven/Cataclysm-DDA
# Fix stop of NPC disassembly job while on other Z-lvl (fix #52008)

#### Summary
Bugfixes "Fix stop of NPC disassembly and crafting jobs while player changes Z-level"


#### Purpose of change
Fixes #52008

Resolves the issue of NPC companions stopping their disassembly activity when player changes Z-level. It most likely will resolve similar issue for crafting activities performed by NPCs

!!! See my post below for newest description of the solution - this one will stay to keep PR history intact !!!


#### Describe the solution
The underlying issue is the fact that no light level calculations are performed for Z-levels other than one player is on. NPCs performing an activity (or trying to start next one scheduled [e.g. next item disassembly]) were cancelling due to not meeting light level requirement (or craft speed requirement derived from light level).

Solution adds a "is_avatar()" check to those light level requirements to make them valid only for PC. NPCs are excluded and therefore can continue their assigned activity on different Z-level.

Solution has one major drawback: NPCs can disassemble in the dark - I am currently looking  for a better way to solve the issue and would gladly accept help/suggestions. Please take a look into alternatives I consider

Solution was based on comment found in #52008. It suggested that similar fix as #46615 might be needed in this case. I investigated and came to the conclusion that it indeed might work. Furthermore, a fact that similar solution was already merged for reading activity made it more probable as a valid approach.



#### Describe alternatives you've considered
Alternatives that solve Z-lvl issue but still forbid NPCs to craft/disassemble in the dark:
1) PREFERRED: Determine if workplace/NPC location is counted "inbounds" the same way terrain is checked for light calculations. If not use the override (to allow crafting while on different z-lvl). If place is "inbounds" it will follow all regular rules and not allow crafting in the dark. Additional advantage is that start of NPC activity has to be when light is sufficient. Once player changes z-lvl and light calculations cease for the workplace it is assumed that light is still present there (it was at the activity start).

I am currently looking into lightmap.cpp and character.cpp (fine_detail_vision_mod function) to somehow make it work. I think I keep failing at proper inbounds implementation.

2) Adding some check if the terrain has current light calculation and use override from original solution only when flag is absent. Result would be same as option 1) but it seems to be less efficient and elegant.


#### Testing
!!! I acknowledge that current solution might be unacceptable due to NPCs crafting in the dark - I still make this PR to give this issue a push and possibly receive a proper advice from someone more experienced with our C++ code !!!
I am fairly certain that it can be reasonably solved if I mange to properly check if NPC is in the area with calculated light or not.

- Compiled game with my code changes, 
- Started a new world and character
- Recruited starting NPC
- Spawned items to disassemble (smartphones) and necessary tools
- Set up proper tool and disassembly zones in unlit basement bathroom (starting shelter)
- Confirm there is no light
- Set up bare-bones basecamp in the basement, built the bulletin board
- Tried disassembly with PC (lack of light) and confirm it is impossible

1) Manual task assignment
- Asked NPC to disassemble items
- Confirmed NPC starts and continues with the job
- Went upstairs  and an overmap tile or two out of the shelter
- Waited 30 min
- Went back downstairs and confirmed that NPC is still busy and disassembled several items

2) Basecamp task
- Tell NPC to stop activity and follow me
- Set up basecamp zone (storage)
- Assigned NPC to basecamp as worker
- Set up Disassembly job priority for NPC (as highest)
- Observe NPC starting and making progress with the task
- Went upstairs, out of shelter, wait
- Went back and confirm NPC is still busy and disassembled several items

3) Notes
- Teleporting a few overmap tiles further stops the activity (separate function in character.cpp does it and I assume it is intended).
- Tests were performed with light on tiles as well; results as expected were the same and activity was performed
- Disassembly by PC in the presence of light was available as expected



#### Additional context
